Add experiments docs (#50)
discord-userdoccers/discord-userdoccers: pages/topics/experiments.mdx

| Field | Type                                    | Description                       |
| ----- | --------------------------------------- | --------------------------------- |
| type  | [filter type hash](#population-filters) | Murmur3 hash of the filter's type |
| value | filter data                             | Explained above in the enum       |

### Bucket Override Object

An override represents a manual setting by Discord staff to grant a resource early or specific access to an experiment. Contrary to what you'd
expect, guild experiments can be and are often overriden and assigned to a user ID.

###### Bucket Override Example

```json
{
  "b": 1,
  "k": ["882680660588904448", "882703776794959873", "859533785225494528", "859533828754505741"]
}
```

###### Bucket Override Structure

| Field | Type             | Description                             |
| ----- | ---------------- | --------------------------------------- |
| b     | integer          | Bucket assigned to these resources      |
| k     | array[snowflake] | Resources granted access to this bucket |

## Endpoints

<RouteHeader method="GET" url="/experiments">
  Get Experiment Assignments
</RouteHeader>

> Returns a fingerprint if not already set, and the [user experiment assignments](#user-experiments) for the fingerprint. Additionally, if `with_guild_experiments` is passed, returns the [guild experiment rollouts](#guild-experiments).

Reviewer: ```suggestion
Returns the [user experiment assignments](#user-experiments) and optionally [guild experiment rollouts](#guild-experiments) for the user/fingerprint.

<Alert type="info">

A fingerprint will only be generated and returned if no authorization or fingerprint is provided in request headers. Fingerprint generation has a rate limit of 3 valid fingerprints per 2 minutes per IP. While fingerprints will still be returned past this rate limit, they will not be usable.

</Alert>

###### Query String Parameters

| Field | Type | Description |
| ----- | ----- | ------------- |
| with_guild_experiments? | boolean | Whether to include guild experiments in the returned data |
```

###### Response Structure

| Field              | Type                                              | Description                                               |
| ------------------ | ------------------------------------------------- | --------------------------------------------------------- |

> | fingerprint?       | string                                            | Returns a randomly generated [fingerprint](#fingerprints) |

Reviewer: ```suggestion
| fingerprint?       | string                                            | A generated [fingerprint](#fingerprints) of the current date and time |
```

# Experiments

Experiments are a form of A/B testing used by Discord in the client- and server-side of their applications to serve
different experiences or behaviours to different users randomly and/or based on location, client version, etc.

> At its core, rollouts are simply a YAML value inputted in the Discord admin panel, but obviously not all of the data in
> there is required by the clients so the actual value most clients see is obfuscated and complex to decipher.

Reviewer: ```suggestion
At their core, rollouts are simply YAML inputted into the Discord admin panel, but not all of the data
is required by the clients so the actual value most clients see is obfuscated and complex to decipher.
```

## Fingerprints

> Even the marketing website uses A/B tests to hook users into the app and so the app needs a unique way to identify the
> person using the website, so it resorts to using "fingerprints". Now this isn't the usual method of analysing a browser's
> unique traits, instead the [Get Experiment Rollouts](#get-experiment-rollouts) endpoint simply generates a new ID for each unauthenticated request
> which the marketing website then consumes.

Reviewer: ```suggestion
Discord needed a way to uniquely identify people using Discord without authentication. Even the marketing website uses A/B tests to hook users into the app, and register/login flows have various experiments, so they resorted to using "fingerprints".
These aren't the usual fingerprints generated by collecting information through the browser, instead they are snowflakes generated by unauthenticated requests to [Get Experiment Rollouts](#get-experiment-rollouts). It is expected that fingerprints are sent in the `X-Fingerprint` header in all subsequent requests to the API until authentication, in order to track A/B tests and allow access to API-locked portions of experiments.
```

A fingerprint is comprised of a snowflake and a random hash value, presumably. It looks like this: `1084179945133187083.JQddgNMmwJPghoBtFmaH7jTmdsw`.

<!-- TODO: link to auth page on this -->

> When signing up, the website then passes the fingerprint and the snowflake part of it is passed internally as the user's new
> ID. This means that theoretically the snowflake's creation time on a user's account represents the first time
> they ever visited the Discord homepage.

Reviewer: ```suggestion
When registering a new account, the fingerprint is passed, and (if valid) is used as the created user's ID. This is done in order to preserve experiments across to the registered user.
Therefore, a user account's creation time theoretically represents the first time they visited Discord.
```

This position is used in conjunction with the rollout populations and filters to figure out what the assigned bucket for the experiments are by simply checking which treatment the population is included in[.](https://gist.github.com/c3f521becd84b923c2778546c79a7a03)

```py
murmurhash3.hash('exp_name:resource_id', signed=False) % 1e4
```

## Data Structures

There are two types of experiments, user experiments and guild experiments. Metadata and human-readable experiment names
are available in the user-facing clients but API objects do not contain such data, except for guild experiments which do
have a human readable name for some of them.

All of the below objects are represented as arrays following the order in which the fields are displayed.

### User Experiments

User experiment data returned by the API is very limited. In contrast to the wide range of data the API provides for
guild rollouts, the only values we can programmatically retrieve from the API are the user/[fingerprint](#fingerprints)'s assigned
bucket for the experiment.

> ###### User Experiment Assignment Structure

Reviewer: ```suggestion
###### User Experiment Structure
```

<Alert type="info">

A fingerprint will only be generated and returned if no authorization or fingerprint is provided in request headers. Fingerprint generation has a rate limit of 3 valid fingerprints per 2 minutes per IP. While fingerprints will still be returned past this rate limit, they will not be usable.

</Alert>

###### Query String Parameters

| Field                   | Type    | Description                                               |
| ----------------------- | ------- | --------------------------------------------------------- |
| with_guild_experiments? | boolean | Whether to include guild experiments in the returned data |

###### Response Structure

| Field              | Type                                              | Description                                                           |
| ------------------ | ------------------------------------------------- | --------------------------------------------------------------------- |
| fingerprint?       | string                                            | A generated [fingerprint](#fingerprints) of the current date and time |

> | assignments        | array[[experiment assignment](#user-experiments)] | The assignment for this fingerprint or user                           |

Reviewer: ```suggestion
| assignments        | array[[experiment assignment](#user-experiments)] | The experiment assignments for this user or fingerprint                           |
```
